packages/yarnpkg-fslib/sources/algorithms/copyPromise.ts

```diff
@@ -86,6 +86,9 @@ async function copyFolder<P1 extends Path, P2 extends Path>(operations: Operatio
 
   const entries = await sourceFs.readdirPromise(source);
 
+  // Readdir isn't necessarily sorted on all systems
+  entries.sort();
+
   await Promise.all(entries.map(async entry => {
```

Comment:
As per our Discord conversation, the sorting here doesn't imply sequential copies, since the mapped promises can terminate out of order.

However, this may be enforced by a synchronous loop instead:
```js
for (const entry of entries.sort()) {
  await copyImpl(operations, lutimes, destinationFs, destinationFs.pathUtils.join(destination, entry), sourceFs, sourceFs.pathUtils.join(source, entry), opts);
}
```